File: pathtree/fullpath.js
// fullpath.js: reconstruct full path for each filehandle from fhparent mapping
// stdin: CSV fh,name,parent
// stdout: CSV fh,top-level unresolved fh,path

var util = require('util');
var fhparent_func = require('./fhparent.func.js')
var fullpath_func = require('./fullpath.func.js')

var PROGRESS = 100000; // interval of progress report, 0 to disable
var progressHeading = 'PROGRESS';
var progressCount = 0;
function incrementProgress() {
  if (PROGRESS > 0 && (++progressCount) % PROGRESS == 0) {
    console.warn(progressHeading, progressCount, process.memoryUsage().rss);
  }
}

var relations = {}; // fh=>{name:name,[parent:parent_fh],[mountpoint:true]}

function processRow(row) {
  var fh = row.fh, parent = row.parent;
  var relation = { name:row.name };
  if (parent == 'MOUNTPOINT') {
    relation.mountpoint = true;
  }
  else {
    relation.parent = parent;
  }
  relations[fh] = relation;
}

var csvParser = fhparent_func.makeCsvParser();
csvParser.on('readable', function(){
  var row;
  while (row = csvParser.read()) {
    processRow(row);
    incrementProgress();
  }
});
csvParser.on('finish', function(){
  writeResults();
});
if (PROGRESS > 0) {
  progressHeading = 'parse';
  progressCount = 0;
}
process.stdin.pipe(csvParser);


var csvStringifier = fullpath_func.makeCsvStringifier();
csvStringifier.pipe(process.stdout);

function constructPath(fh, tail) {
  var relation = relations[fh];
  if (!relation) { // unresolved
    return { 'top':fh, 'path':tail };
  }

  tail = tail ? '/' + tail : '';

  if (relation.mountpoint) {
    return { 'top':'', 'path':relation.name + tail };
  }

  return constructPath(relation.parent, relation.name + tail);
}

function writeResults() {
  if (PROGRESS > 0) {
    progressHeading = 'output';
    progressCount = 0;
  }
  Object.keys(relations).forEach(function(fh){
    var p = constructPath(fh);
    p.fh = fh;
    csvStringifier.write(p);
    incrementProgress()
  });
}

File: pathtree/fullpath.cc
// fullpath.cc: reconstruct full path for each filehandle from fhparent mapping
// compile: g++ --std=c++0x -o fullpath fullpath.cc
// run: cat x.fhparent | tr ',' '\t' | ./fullpath > x.fullpath
// stdin: CSV fh,name,parent
// stdout: CSV fh,top-level unresolved fh,path

#include <tuple>
#include <string>
#include <unordered_map>
#include <iostream>
#include <fstream>

const int PROGRESS = 100000; // interval of progress report, 0 to disable
std::string progressHeading = "PROGRESS";
int progressCount = 0;

// http://stackoverflow.com/a/19770392/3729203
void process_mem_usage(double& vm_usage, double& resident_set)
{
    vm_usage     = 0.0;
    resident_set = 0.0;

    // the two fields we want
    unsigned long vsize;
    long rss;
    {
        std::string ignore;
        std::ifstream ifs("/proc/self/stat", std::ios_base::in);
        ifs >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore
                >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore
                >> ignore >> ignore >> vsize >> rss;
    }

    long page_size_kb = sysconf(_SC_PAGE_SIZE) / 1024; // in case x86-64 is configured to use 2MB pages
    vm_usage = vsize / 1024.0;
    resident_set = rss * page_size_kb;
}

void
incrementProgress()
{
  double vm_usage, resident_set;
  if (PROGRESS > 0 && (++progressCount) % PROGRESS == 0) {
    process_mem_usage(vm_usage, resident_set);
    std::cerr << progressHeading << " " << progressCount << " " << resident_set << std::endl;
  }
}

struct Relation
{
  std::string name;
  std::string parent;
  bool mountpoint;
};
typedef std::unordered_map<std::string, Relation> Relations; // fh=>relation

Relations
parseInput(std::istream& input)
{
  Relations relations;
  progressHeading = "parse";
  progressCount = 0;

  std::string fh, name, parent;
  while (input >> fh >> name >> parent) {
    Relation& relation = relations[fh];
    relation.name = name;
    relation.mountpoint = parent == "MOUNTPOINT";
    if (!relation.mountpoint) {
      relation.parent = parent;
    }
    incrementProgress();
  }
  return relations;
}

std::tuple<std::string, std::string> // top, path
constructPath(const Relations& relations, std::string fh, std::string tail = "")
{
  auto relationIt = relations.find(fh);
  if (relationIt == relations.end()) { // unresolved
    return std::forward_as_tuple(fh, tail);
  }
  const Relation& relation = relationIt->second;

  std::string newTail = relation.name;
  if (!tail.empty()) {
    newTail.push_back('/');
    newTail.append(tail);
  }

  if (relation.mountpoint) {
    return std::forward_as_tuple("", newTail);
  }

  return constructPath(relations, relation.parent, newTail);
}

void
writeResults(const Relations& relations, std::ostream& output)
{
  progressHeading = "output";
  progressCount = 0;

  for (const auto& pair : relations) {
    const std::string& fh = pair.first;
    const Relation& relation = pair.second;
    std::string top, path;
    std::tie(top, path) = constructPath(relations, fh);

    output << fh << ',' << top << ',' << path << '\n';
    incrementProgress();
  }
}

int
main()
{
  Relations relations = parseInput(std::cin);
  writeResults(relations, std::cout);
  return 0;
}
